src/dbnode/storage/index.go

```diff
@@ -1436,3 +1440,230 @@ func (shards dbShards) IDs() []uint32 {
+	// Capture start before needing to acquire lock.
+	start := i.nowFn()
+
+	i.state.RLock()
+	if !i.isOpenWithRLock() {
+		i.state.RUnlock()
+		return nil, errDbIndexUnableToQueryClosed
+	}
+
+	// Track this as an inflight query that needs to finish
+	// when the index is closed.
+	i.queriesWg.Add(1)
+	defer i.queriesWg.Done()
+
+	// Enact overrides for query options
+	opts.QueryOptions = i.overriddenOptsForQueryWithRLock(opts.QueryOptions)
+	timeout := i.timeoutForQueryWithRLock(ctx)
+
+	// Retrieve blocks to query, then we can release lock
+	// NB(r): Important not to block ticking, and other tasks by
+	// holding the RLock during a query.
+	blocks, err := i.blocksForQueryWithRLock(xtime.NewRanges(xtime.Range{
+		Start: opts.StartInclusive,
+		End:   opts.EndExclusive,
+	}))
+
+	// Can now release the lock and execute the query without holding the lock.
+	i.state.RUnlock()
+
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		deadline = start.Add(timeout)
+		wg       sync.WaitGroup
+
+		// Results contains all concurrent mutable state below.
+		results = struct {
+			sync.Mutex
+			multiErr   xerrors.MultiError
+			merged     index.AggregateResults
+			exhaustive bool
+			returned   bool
+		}{
+			merged:     nil,
+			exhaustive: true,
+			returned:   false,
+		}
+	)
+	defer func() {
+		// Ensure that during early error returns we let any aborted
+		// goroutines know not to try to modify/edit the result any longer.
+		results.Lock()
+		results.returned = true
+		results.Unlock()
+	}()
+
+	execBlockQuery := func(block index.Block) {
+		aggregateResults := i.aggregateResultsPool.Get()
+		aggregateResults.Reset(i.nsMetadata.ID())
+
+		err := block.AggregateQuery(query, opts, aggregateResults)
+		if err == index.ErrUnableToQueryBlockClosed {
+			// NB(r): Because we query this block outside of the results lock, it's
+			// possible this block may get closed if it slides out of retention, in
+			// that case those results are no longer considered valid and outside of
+			// retention regardless, so this is a non-issue.
+			err = nil
+		}
+
+		var mergedResult bool
+		results.Lock()
+		defer func() {
+			results.Unlock()
+			if mergedResult {
+				// Only finalize this result if we merged it into another.
+				aggregateResults.Finalize()
+			}
+		}()
+
+		if results.returned {
+			// If already returned then we early cancelled, don't add any
+			// further results or errors since caller already has a result.
+			return
+		}
+
+		if err != nil {
+			results.multiErr = results.multiErr.Add(err)
+			return
+		}
+
+		if results.merged == nil {
+			// Return results to pool at end of request.
+			ctx.RegisterFinalizer(aggregateResults)
+			// No merged results yet, use this as the first to merge into.
+			results.merged = aggregateResults
+		} else {
+			// Append the block results.
+			mergedResult = true
+			for _, entry := range aggregateResults.Map().Iter() {
+				// Append to merged results.
+				id, tags := entry.Key(), entry.Value()
```

Comment:
why is the ID needed in aggregation results?

Reply:
Probably a bad name; `id` here represents unique tag names